io/shared/src/main/scala/fs2/io/net/tls/TLSSocket.scala

```diff
@@ -38,9 +38,19 @@ sealed trait TLSSocket[F[_]] extends Socket[F] with TLSSocketPlatform[F] {
   def session: F[SSLSession]
 
   /** Provides access to the current application protocol that has been negotiated.
+    *
+    * Raises a [[NoSuchElementException]] if it has not yet been determined if application protocols might
+    * be used for this connection. See [[applicationProtocolOption]] for a safer option.
+    *
+    * @see https://discord.com/channels/632277896739946517/632310980449402880/1100649913223745597
     */
   def applicationProtocol: F[String]
 
+  /** Provides access to the current application protocol that has been negotiated.
+    *
+    * Returns [[None]] if it has not yet been determined if application protocols might be used for this connection.
```

Comment:
Maybe we should include the bit about the empty `String`? Actually, this _almost_ deserves its own ADT but I'm not sure if it's worth it 🤔 or, maybe that case should be mapped to `None` as well? I'm not sure if there is a meaningful distinction.
> null if it has not yet been determined if application protocols might be used for this connection, an empty String if application protocols values will not be used, or a non-empty application protocol String if a value was successfully negotiated.